### PR TITLE
Update Shuffle Function to Swap-or-not

### PR DESCRIPTION
**Pull request by cyberbono3**
[Resolves] #1529

---

# Description

1. I have implemented shuffle algorithm according to [spec](https://github.com/ethereum/eth2.0-specs/pull/576)

2. I ran bazel test and majority of tests failed with timeout error. 

3.  I think with high probability my implementation is not correct.

I would appreciate your review and feedback.

### CODE REVIEW

**Comment by prestonvanloon** [comment]
Many tests failing, can you please take a look?

File: beacon-chain/utils/shuffle.go [hunk 1]
```diff
@@ -2,57 +2,52 @@
...
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/prysmaticlabs/prysm/shared/hashutil"
-	"github.com/prysmaticlabs/prysm/shared/params"
+	//"github.com/prysmaticlabs/prysm/shared/params"
 )
 
 // ShuffleIndices returns a list of pseudorandomly sampled
 // indices. This is used to shuffle validators on ETH2.0 beacon chain.
 func ShuffleIndices(seed common.Hash, indicesList []uint64) ([]uint64, error) {
```
Comment (nisdas) [inline beacon-chain/utils/shuffle.go]: This is confusing this doesnt match the spec at all , this is the function in the spec
```py
def get_permuted_index(index: int, list_size: int, seed: Bytes32) -> int:
```
There also no assertions from the spec 
```py
    assert index < list_size
    assert list_size <= 2**40
```

File: beacon-chain/utils/shuffle.go [hunk 2]
```diff
@@ -2,57 +2,52 @@
...
-	// Each entropy is consumed from the seed in randBytes chunks.
-	randBytes := params.BeaconConfig().RandBytes
+	var hashBytes []byte
+	bs1 := make([]byte, 8)
+	bs4 := make([]byte, 8)
+	listSize := len(indicesList)
+	num := int(math.Floor(float64((listSize + 255) / 256)))
+	powersOfTwo := []uint64{1, 2, 4, 8, 16, 32, 64, 128}
```
Comment (nisdas) [inline beacon-chain/utils/shuffle.go]: Please use the `PowerOf2`function in `shared/mathutils` for this

File: beacon-chain/utils/shuffle_test.go [hunk 3]
```diff
@@ -47,12 +47,14 @@ func TestShuffleIndices(t *testing.T) {
...
 		t.Errorf("2 shuffled lists shouldn't be equal")
 	}
-	if !reflect.DeepEqual(list1, []uint64{5, 4, 9, 6, 7, 3, 0, 1, 8, 2}) {
-		t.Errorf("list 1 was incorrectly shuffled")
-	}
-	if !reflect.DeepEqual(list2, []uint64{9, 0, 1, 5, 3, 2, 4, 7, 8, 6}) {
-		t.Errorf("list 2 was incorrectly shuffled")
-	}
+
+	// if !reflect.DeepEqual(list1, []uint64{3,3,0,3,0,3,0,2,0,3}) {
```
Comment (nisdas) [inline beacon-chain/utils/shuffle_test.go]: dont comment out code, either check it in or remove it.

**Comment by prestonvanloon** [comment]
This implementation might be a helpful reference or library to use: https://github.com/protolambda/eth2-shuffle

**Comment by cyberbono3** [comment]
thanks for your feedback, guys !
@nisdas  i am a bit confused, what  function should I implement [get_shuffling](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#get_shuffling)  or [get_permuted_index](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#get_permuted_index) ?

**Comment by prestonvanloon** [comment]
Please fix tests

File: beacon-chain/utils/shuffle.go [hunk 4]
```diff
@@ -2,61 +2,67 @@
...
 
 import (
+	"encoding/binary"
 	"errors"
+	"math"
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/prysmaticlabs/prysm/shared/hashutil"
-	"github.com/prysmaticlabs/prysm/shared/params"
+	//"github.com/prysmaticlabs/prysm/shared/params"
```
Comment (nisdas) [inline beacon-chain/utils/shuffle.go]: Please remove this, if you are not using. dont comment it out

**Comment by prestonvanloon** [comment]
Taking a look at this to see what I can do to help

**Comment by prestonvanloon** [comment]
Blocked by testnet feature freeze

**Pull request closed without merging**